### DISCUSSION TITLE
`EndMarker` and how to handle it

### DISCUSSION THREAD

**Discussion by KaranAhlawat** [tree-sitter/tree-sitter-scala#313] [Q&A]
So, for a while, I've been trying to bend `EndMarker`s to my will, since I opened #275. But, unfortunately, it has escaped me how we can group them. I think, looking at the EBNF grammar, it's not really possible to group them with their corresponding openings.

At least not if we implement the tree-sitter grammar according to the EBNF. But, even if we do deviate from it, we still run into issues because we can't really figure out which statement an `end` marker might belong to, without the rest of the parse tree.

So, my question is, should I just fall back to using regexp to determine the correct indentation for a `end` marker? Or is there something we can do on the parser level for this?

**Comment by susliko** [comment]
Deviations from the ebnf are okay.
I believe, that we can handle them with tree-sitter. I did some digging and was able to parse a majority of end marker use cases, but some tests for other cases broke, and I haven't spent enough time to fix them

**Comment by KaranAhlawat** [comment]
I haven't been able to make them optional *and* parse them correctly. Also, wasn't able to figure out how to capture those correctly for `given_expressions`. If you guys have any pointers, I'd love to go back to this once again and try to fix it.
